accounting/src/delta/delta_data_collection/tests/mod.rs

```diff
@@ -0,0 +1,322 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::*;
+
+mod delta_delta_delta_delta_tests;
+mod delta_delta_delta_tests;
+mod delta_delta_undo_tests;
+mod delta_delta_undo_undo_tests;
+
+#[test]
+#[rustfmt::skip]
+fn test_combine_deltas() {
+    use DataDelta::{Create, Delete, Modify};
+
+    assert_eq!(combine_delta_data(&Create('a'), Create('b')),      Err(Error::DeltaDataCreatedMultipleTimes));
+    assert_eq!(combine_delta_data(&Create('a'), Modify('a', 'b')), Ok(Some(Create('b'))));
+    assert_eq!(combine_delta_data(&Create('a'), Delete('a')),      Ok(None));
+
+    assert_eq!(combine_delta_data(&Modify('a', 'b'), Create('c')),      Err(Error::DeltaDataCreatedMultipleTimes));
+    assert_eq!(combine_delta_data(&Modify('a', 'b'), Modify('c', 'd')), Ok(Some(Modify('a', 'd'))));
+    assert_eq!(combine_delta_data(&Modify('a', 'b'), Modify('b', 'a')), Ok(None));
+    assert_eq!(combine_delta_data(&Modify('a', 'b'), Delete('b')),      Ok(Some(Delete('a'))));
+
+    assert_eq!(combine_delta_data(&Delete('a'), Create('a')),      Ok(None));
+    assert_eq!(combine_delta_data(&Delete('a'), Create('b')),      Ok(Some(Modify('a', 'b'))));
+    assert_eq!(combine_delta_data(&Delete('a'), Modify('b', 'c')), Err(Error::DeltaDataModifyAfterDelete));
+    assert_eq!(combine_delta_data(&Delete('a'), Delete('b')),      Err(Error::DeltaDataDeletedMultipleTimes));
+}
+
+#[test]
+fn create_delta_undo_roundtrip() {
+    let check_undo = |delta1, delta2: DataDelta<char>| {
+        let undo = create_undo_delta(delta2.clone());
+        let combine_result = combine_delta_data(&delta1, delta2).unwrap().unwrap();
+        let undo_result = combine_delta_data(&combine_result, undo.0).unwrap().unwrap();
+
+        assert_eq!(delta1, undo_result);
+    };
+
+    check_undo(DataDelta::Create('a'), DataDelta::Modify('a', 'b'));
+    check_undo(DataDelta::Modify('a', 'b'), DataDelta::Delete('b'));
+    check_undo(DataDelta::Modify('a', 'b'), DataDelta::Modify('b', 'c'));
+    check_undo(DataDelta::Delete('a'), DataDelta::Create('b'));
+}
+
+// Same as `create_delta_undo_roundtrip` but for deltas that produce No-op
+#[test]
+fn create_delta_undo_noop_roundtrip() {
+    let check_undo = |delta1, delta2: DataDelta<char>| {
+        let undo = create_undo_delta(delta2.clone());
+        let combine_result = combine_delta_data(&delta1, delta2).unwrap();
+        assert!(combine_result.is_none());
+        assert_eq!(delta1, undo.0);
+    };
+
+    check_undo(DataDelta::Modify('a', 'b'), DataDelta::Modify('b', 'a'));
+    check_undo(DataDelta::Create('a'), DataDelta::Delete('a'));
+    check_undo(DataDelta::Delete('a'), DataDelta::Create('a'));
+}
+
+#[test]
+fn merge_collections_and_undo() {
```

Comment:
I'm OK with having this test, but do we have a similar test at the combine level? Or is this just at the collection merge level?

Reply:
I'm talking specifically about the undos, btw. It seems that this test is special because it tests the undo cases.